terminal.c: delete 10% of scrollback when full

danielBelenky opened the pull request:
Instead of deleting a single line at a time when the scrollback is full
in terminal mode, will delete 10% of the lines.

Is a partial solution (workaround?) to https://github.com/neovim/neovim/issues/8811

danielBelenky committed:
terminal.c: delete 10% of scrollback when full

Instead of deleting a single line at a time when the scrollback is full
in terminal mode, will delete 10% of the lines.

justinmk commented on src/nvim/terminal.c:
```diff
@@ -1192,9 +1192,14 @@ static void refresh_scrollback(Terminal *term, buf_T *buf)
+        ml_delete(1, false);
+        deleted_lines(1, 1);
+      }
     }
```
maybe this if-block can go out of the while-loop, and the while-loop can become a for-loop. And then `appended_lines` could also be moved out of the loop.

Commit:
terminal.c: optimize refresh_scollback function

- Move condition block that checks if buffer is full out of the loop that
  syncs lines from libvterm to the current memory line.
- Move call to updated_lines from out of the loop that syncs lines from
  libvterm to the current memory line.
- Added 10% to scrollback calculation to allow deletion of lines in
  chunks.
- When scrollback buffer gets full, delete 10% of the top rows instead
  of deleting one line at a time.
- Replaced while loop that appends terminal lines to the current
  buffer with a for loop.

bfredl commented:
This means the scrollback is effectively only 90% what the user requested. Wouldn't it be less disturbing if nvim instead collected lines _in addition_ (which could be less than 10% for a large scrollback), so a larger delete could be done _down_ to the user requested size?

danielBelenky commented:
> This means the scrollback is effectively only 90% what the user requested.

Right. Maybe it is not the best idea to blindly delete 10% of the scrollback.

> Wouldn't it be less disturbing if nvim instead collected lines in addition (which could be less than 10% for a large scrollback), so a larger delete could be done down to the user requested size?

I'm not sure I fully understand your suggestion... You want to delete 10% but only up to the user's definition?

bfredl commented:
@danielBelenky I mean we should let the scrollback buffer grow slightly larger than the users requested size, so we can do a large delete, without going below the requested size. If the user requested 1000 lines, it is better to delete from 1100 down to 1000, than from 1000 down to 900.

justinmk commented:
That seems good. So just set `sb_size` to 'scrollback' + 10% extra at https://github.com/neovim/neovim/blob/6ec9d790e11b009a8c002987bdbcb6250848aad5/src/nvim/terminal.c#L252-L253

something like this:

```c
int scbk_plus_10pct = curbuf->b_p_scbk
                      + (int)((double)curbuf->b_p_scbk * .10);
rv->sb_size = curbuf->b_p_scbk < 0
                ? SB_MAX : (size_t)MAX(1, scbk_plus_10pct);
```

Note: in the case of `SB_MAX`, don't bother with 10% extra.

justinmk commented on src/nvim/terminal.c:
```diff
@@ -1184,23 +1184,31 @@ static void on_scrollback_option_changed(Terminal *term, buf_T *buf)
+    int sb_delete = (int)term->sb_size / 10;
+    for(int i = 1; i <= sb_delete; ++i, --term->sb_current) {
+        ml_delete(i, false);
+        xfree(term->sb_buffer[term->sb_current]);
```
I'm not sure that `xfree` is wanted here. Note that `on_scrollback_option_changed` does that because when the actual 'scrollback' option changes we want libvterm's internal scrollback to change. But not here.

Here (in `refresh_scrollback`) we're only updating the vim buffer.

danielBelenky replied:
> I'm not sure that xfree is wanted here.

I see... but the memory is still being allocated to libvterm's internal scrollback. Doesn't it bother us? Or `on_scrollback_option_changed` should take care of that?

justinmk replied:
> but the memory is still being allocated to libvterm's internal scrollback

yes but that allocation isn't done by `refresh_scrollback` , it's done elsewhere, and also freed elsewhere, just as before this PR.

danielBelenky replied:
> yes but that allocation isn't done by refresh_scrollback , it's done elsewhere, and also freed elsewhere (terminal_destroy), just as before this PR.

Thanks for explaining

danielBelenky commented:
I see that the PR fail one of the tests:

```
Passed in:
(table) {
 *[1] = '2: foobar'
  [2] = '3: foobar'
  [3] = '4: foobar'
  [4] = '5: foobar'
  [5] = '6: foobar'
  [6] = '7: foobar'
  [7] = '8: foobar'
  [8] = '9: foobar'
  [9] = '61: foobar'
  [10] = '95: foobar'
  [11] = '96: foobar'
  [12] = ''
  [13] = '[Process exited 0]' }
Expected:
(table) {
 *[1] = '86: foobar'
  [2] = '87: foobar'
  [3] = '88: foobar'
  [4] = '89: foobar'
  [5] = '90: foobar'
  [6] = '91: foobar'
  [7] = '92: foobar'
  [8] = '93: foobar'
  [9] = '94: foobar'
  [10] = '95: foobar'
  [11] = '96: foobar'
  [12] = ''
  [13] = '[Process exited 0]' }
stack traceback:
	C:/projects/neovim/test/functional\terminal\edit_spec.lua:67: in function <C:/projects/neovim/test/functional\terminal\edit_spec.lua:34>
```

I'll try to fix the PR.

danielBelenky commented on src/nvim/terminal.c:
```diff
@@ -1186,21 +1189,29 @@ static void refresh_scrollback(Terminal *term, buf_T *buf)
-  while (term->sb_pending > 0) {
+  if (((int)buf->b_ml.ml_line_count - height) >= (int)term->sb_size) {
+    // scrollback buffer is full - delete 10% from the top
+    int sb_delete = (int)term->sb_size / 10;
```
Thinking about this now, it's wrong to just calculate 10% out of `term->sb_size` as it now includes the 10% delta.
Since I'll need to know the delta that was appended to `term->sb_size` when calculating how many rows can be deleted, I was thinking of either declaring `static const int SCBK_DELTA = 1.1` and then taking it into account when calculating `sb_delete`. Another option I've thought of is to exit the loop that executes `ml_delete` when `buf->b_ml.ml_line_count - height <= term->sb_size`

@justinmk @bfredl WDYT?

justinmk replied:
@danielBelenky It's wrong, but it doesn't matter right now. What matters is getting it to actually work, without failing the ASan build (the first build listed in Travis CI).

justinmk commented:
> exit the loop that executes `ml_delete` when `buf->b_ml.ml_line_count - height <= term->sb_size`

Makes sense to me.

justinmk commented:
Still want something like this probably, but needs tests. And I think https://github.com/neovim/neovim/pull/39773 does something like this.
